File: android/lib/crypt/aes_gcm_256_pbkdf2_string_encryption.dart
// Source: https://github.com/java-crypto/cross_platform_crypto/tree/main/AesGcm256Pbkdf2StringEncryption

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import "package:pointycastle/export.dart";

String aesGcmPbkdf2EncryptToBase64(String password, String plaintext) {
  var plaintextUint8 = createUint8ListFromString(plaintext);
  var passphrase =  createUint8ListFromString(password);
  const pbkdf2Iterations = 15000;
  final salt = generateSalt32Byte();
  KeyDerivator derivator = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64));
  Pbkdf2Parameters params = Pbkdf2Parameters(salt, pbkdf2Iterations, 32);
  derivator.init(params);
  final key = derivator.process(passphrase);
  final nonce = generateRandomNonce();
  final cipher = GCMBlockCipher(AESEngine());
  var aeadParameters = AEADParameters(KeyParameter(key), 128, nonce, Uint8List(0));
  cipher.init(true, aeadParameters);
  var ciphertextWithTag = cipher.process(plaintextUint8);
  var ciphertextWithTagLength = ciphertextWithTag.lengthInBytes;
  var ciphertextLength = ciphertextWithTagLength - 16; // 16 bytes = 128 bit tag length
  var ciphertext = Uint8List.sublistView(ciphertextWithTag, 0, ciphertextLength);
  var gcmTag = Uint8List.sublistView(ciphertextWithTag, ciphertextLength, ciphertextWithTagLength);
  final saltBase64 = base64Encoding(salt);
  final nonceBase64 = base64Encoding(nonce);
  final ciphertextBase64 = base64Encoding(ciphertext);
  final gcmTagBase64 = base64Encoding(gcmTag);
  return '$saltBase64:$nonceBase64:$ciphertextBase64:$gcmTagBase64';
}

String aesGcmPbkdf2DecryptFromBase64(String password, String data) {
  var parts = data.split(':');
  var salt = base64Decoding(parts[0]);
  var nonce = base64Decoding(parts[1]);
  var ciphertext = base64Decoding(parts[2]);
  var gcmTag = base64Decoding(parts[3]);
  var bb = BytesBuilder();
  bb.add(ciphertext);
  bb.add(gcmTag);
  var ciphertextWithTag = bb.toBytes();
  var passphrase =  createUint8ListFromString(password);
  const pbkdf2Iterations = 15000;
  KeyDerivator derivator = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64));
  Pbkdf2Parameters params = Pbkdf2Parameters(salt, pbkdf2Iterations, 32);
  derivator.init(params);
  final key = derivator.process(passphrase);
  final cipher = GCMBlockCipher(AESEngine());
  var aeadParameters = AEADParameters(KeyParameter(key), 128, nonce, Uint8List(0));
  cipher.init(false, aeadParameters);
  return String.fromCharCodes(cipher.process(ciphertextWithTag));
}

Uint8List generateSalt32Byte() {
  final sGen = Random.secure();
  final seed =
  Uint8List.fromList(List.generate(32, (n) => sGen.nextInt(255)));
  SecureRandom sec = SecureRandom("Fortuna")..seed(KeyParameter(seed));
  return sec.nextBytes(32);
}

Uint8List generateRandomNonce() {
  final sGen = Random.secure();
  final seed =
  Uint8List.fromList(List.generate(32, (n) => sGen.nextInt(255)));
  SecureRandom sec = SecureRandom("Fortuna")..seed(KeyParameter(seed));
  return sec.nextBytes(12);
}

Uint8List createUint8ListFromString(String s) {
  var ret = Uint8List(s.length);
  for (var i = 0; i < s.length; i++) {
    ret[i] = s.codeUnitAt(i);
  }
  return ret;
}

String base64Encoding(Uint8List input) {
  return base64.encode(input);
}

Uint8List base64Decoding(String input) {
  return base64.decode(input);
}

File: linux/crypt/AesGcm256Pbkdf2StringEncryption.go
// Source: https://github.com/java-crypto/cross_platform_crypto/tree/main/AesGcm256Pbkdf2StringEncryption

package crypt

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
  "encoding/base64"
  "io"
	"strings"
	"golang.org/x/crypto/pbkdf2"
)

func AesGcmPbkdf2EncryptToBase64(passphrase string, data string)(string) {
  PBKDF2_ITERATIONS := int(15000)
  salt := []byte(GenerateSalt32Byte())
  // derive key
  key := []byte(pbkdf2.Key([]byte(passphrase), salt, PBKDF2_ITERATIONS, 32, sha256.New))
  // encrypt
  plaintext := []byte(data)
  nonce := []byte(GenerateRandomNonce())
	block, err := aes.NewCipher(key)
	if err != nil {
		panic(err.Error())
	}
	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		panic(err.Error())
	}
  ciphertext := aesGCM.Seal(nil, nonce, plaintext, nil)
  // ciphertext hold the ciphertext | gcmTag
  ciphertextWithTagLength := int(len(ciphertext))
  ciphertextWithoutTag, gcmTag := ciphertext[:(ciphertextWithTagLength-16)], ciphertext[(ciphertextWithTagLength-16):]
  ciphertextWithoutTagBase64 := string(Base64Encoding(ciphertextWithoutTag))
  saltbase64 := string(Base64Encoding(salt))
  nonceBase64 := string(Base64Encoding(nonce))
  gcmTagBase64 := string(Base64Encoding(gcmTag))
  ciphertextCompleteBase64 := string(saltbase64 + ":" + nonceBase64 + ":" + ciphertextWithoutTagBase64 + ":" + gcmTagBase64) 
  return ciphertextCompleteBase64
}

func AesGcmPbkdf2DecryptFromBase64(passphrase string, ciphertextCompleteBase64 string)(string) {
  PBKDF2_ITERATIONS := int(15000)
  data := strings.Split(ciphertextCompleteBase64, ":") 
  salt := []byte(Base64Decoding(data[0]))
  nonce := []byte(Base64Decoding(data[1]))
  ciphertext := []byte(Base64Decoding(data[2]))
  gcmTag := []byte(Base64Decoding(data[3]))
  ciphertextCompleteLength := int(len(ciphertext) + len(gcmTag))
  ciphertextComplete := make([]byte, ciphertextCompleteLength)
  ciphertextComplete = append(ciphertext[:], gcmTag[:]...)
  // derive key
  key := []byte(pbkdf2.Key([]byte(passphrase), salt, PBKDF2_ITERATIONS, 32, sha256.New))
  // decrypt
  block, err := aes.NewCipher(key)
	if err != nil {
		panic(err.Error())
	}
	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		panic(err.Error())
	}
  plaintext, err := aesGCM.Open(nil, nonce, ciphertextComplete, nil)
	if err != nil {
		panic(err.Error())
	}
  return string(plaintext)
}

func GenerateSalt32Byte()([]byte) {
  salt := make([]byte, 32)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
	  panic(err.Error())
	}
  return salt
}

func GenerateRandomNonce()([]byte) {
  nonce := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
	  panic(err.Error())
	}
  return nonce
}

func Base64Encoding(input []byte)(string) {
  return base64.StdEncoding.EncodeToString(input)
}

func Base64Decoding(input string)([]byte) {
  data, err := base64.StdEncoding.DecodeString(input)
	if err != nil {
		return data
	}
  return data
}
